### ISSUE TITLE
HuggingFace Transformers for SciBERT Fine Tuning

### ISSUE THREAD

**Issue by ivalexander13** [arjunchandran1999/NLPChemExtractor#26]
After looking through example code for relation extraction tasks, writing something to do this task is too monumental. Now we're trying to do NER with HuggingFace, with the advantage (over AllenNLP) of having full control and transparency of the code. 

## Current Actions: NER on Transformers
Looking at example NER code to find a good base skeleton code. Then study AllenNLP's NER design to see any specific fine-tuning steps that we might need to replicate. Finally, begin writing code that fits our needs.

**Comment by ivalexander13** [comment]
Code for NER has been implemented with default Trainer class. Rough hyperparameter tuning has been done and results are logged in this [Google Sheet](https://docs.google.com/spreadsheets/d/1jolvSI9tCqHZqBMtX1MAUjht2WuXyl_uFauhbvHMUtQ/edit#gid=0). The hyperparameters tuned are:
- learning rate: 2e-5
- batch size: 1
- epoch #: tbd (probably 11?)

**To do**:
- Include offset mapping to account for the one-to-many encoding of words.
- Figure out a way to better tune the training loop (how to dig into huggingface's model() function? How to set weights and loss functions?)
- Huggingface for RE: find a good example that isn't too convoluted.

**Comment by ivalexander13** [comment]
To-do Regularizations:
- Early stopping ([huggingface docs](https://huggingface.co/transformers/main_classes/callback.html#transformers.EarlyStoppingCallback)) ([reading](https://machinelearningmastery.com/early-stopping-to-avoid-overtraining-neural-network-models/))
- dropouts ([reading](https://machinelearningmastery.com/dropout-for-regularizing-deep-neural-networks/))

Links:
- Subclassing Trainer to override methods ([hf-docs](https://huggingface.co/transformers/main_classes/trainer.html))
- More discussion about NN design (huggingface, TF, keras) ([tds](https://towardsdatascience.com/hugging-face-transformers-fine-tuning-distilbert-for-binary-classification-tasks-490f1d192379))
